PR: Add system-specific shortcut names on macOS to the Shortcuts summary dialog

leportella opened the pull request:
### Pull Request Checklist

* [x] Read and followed this repo's [Contributing Guidelines](https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md)
* [x] Based this PR on the latest version of the correct branch (master or 3.x)
* [x] Followed [PEP 8](https://www.python.org/dev/peps/pep-0008/) code style
* [x] Ensured this pull request hasn't eliminated unrelated blank lines/spaces,
      modified the ``spyder/defaults`` directory, or added new icons/assets
* [ ] Wrote at least one-line docstrings following PEP 257for any new functions
* [x] Added at least one unit test covering the changes, if at all possible
* [x] Described the changes and the motivation for them below
* [x] Noted what issue(s) this pull request resolves, creating one if needed
* [x] Included a screenshot, if this PR makes any visible changes to the UI


## Description of Changes

I add a system check to see if the system is a MacOS and change the commands `Ctrl` for
 `Cmd` and `Alt` for `Option`.

### Issue(s) Resolved

Fixes #7146


### Screenshot

<img width="1440" alt="screen shot 2018-11-08 at 23 25 48" src="https://user-images.githubusercontent.com/7255297/48237427-b1678900-e3ad-11e8-94dd-2ec741a9dcb1.png">

jnsebgosselin commented:
Hello, thank you very much for you contribution! I'll try to take a look at it this weekend.

CAM-Gerlach commented:
@leportella Does <kbd>Meta</kbd> already get replaced with <kbd>Control</kbd> on macOS? Also, what about <kbd>Backspace</kbd> to <kbd>Delete</kbd>? Do we want to fix these at some lower level so they also display correctly in the menus, keyboard shortcuts manager, etc (or do they already)? Thanks!

jnsebgosselin commented on spyder/widgets/shortcutssummary.py:
```diff
@@ -106,6 +107,11 @@ def __init__(self, parent=None):
                 label_name = QLabel(name.capitalize().replace('_', ' '))
                 label_name.setFont(font_names)
+
+                if sys.platform.startswith('darwin'):
```
If possible, I think it would be better to let Qt do the translation for us instead of doing it ourselves. So, instead of

```
if sys.platform.startswith('darwin'):
    keystr = keystr.replace('Ctrl', 
                                        'Cmd').replace('Alt', 'Option')
```

 could you please test if something like this would work on MacOS to solve this issue :

`keystr  = QKeySequence(keystr).toString(QKeySequence.NativeText)`

You'll need to add this import:

`from qtpy.QtGui import QKeySequence`

CAM-Gerlach replied:
You could also eliminate the platform check since that would be handled automatically.

jnsebgosselin replied:
hehe looking at her profile, I think she's more than capable to figure the details by herself :P

leportella replied:
Hi! I am new to this and I followed what was on the issue. I'll change this as you suggested :)

CAM-Gerlach commented on spyder/widgets/shortcutssummary.py:
```diff
@@ -106,6 +107,11 @@ def __init__(self, parent=None):
+
+                if sys.platform.startswith('darwin'):
+                    keystr = keystr.replace('Ctrl',
+                                            'Cmd').replace('Alt', 'Option')
```
Does <kbd>Meta</kbd> already get replaced with <kbd>Control</kbd> on macOS? Also, what about <kbd>Backspace</kbd> to <kbd>Delete</kbd>? Presumably, @jnsebgosselin 's suggestion would also resolve this.

CAM-Gerlach replied:
But presuming it does, you'd want to include examples of each in your test.

CAM-Gerlach commented on spyder/widgets/shortcutssummary.py:
```diff
@@ -8,14 +8,15 @@
 
 
 # Standard library imports
+import sys
```
This presumably wouldn't be needed if you eliminate the manual platform check and let Qt do it instead.

leportella commented:
@CAM-Gerlach I looked for tests regarding the `ShortcutsSummaryDialog` class. Did I got it wrong or would it be the right thing to create it?

CAM-Gerlach commented:
@CAM-Gerlach You're quiet correct; you'll need to create ``spyder/widgets/tests/test_shortcutsummary.py``. You can use the first ~20 lines of e.g. ``test_arraybuilder.py`` in that directory as a template for how we typically structure the headers of our test files, and you'll probably want to make instantiating a shortcutsummary widget a fixture and pass that to the test function.

CAM-Gerlach commented:
> And please don't think I don't need help! hehehehhe I really do, but I really what to contribute :)

I have a habit of over-explaining things and being too verbose with other new contributors, so I particularly didn't want to do that here with someone having as much Python experience as yourself. Please let us know if you'd like guidance on something and we'd be happy to help.

And we really appreciate yours! As @jnsebgosselin we need all the help we can get, since we are a small team in our spare time serving hundreds of thousands of users across the world, and especially as a Mac developer your contributions are uniquely valuable.

leportella commented:
I instantiated the class `ShortcutsSummaryDialog` but I am not sure how to test the widgets added. I navigated a little bit on the object, but I couldn't find were the final texts are. Could you help me, please? @CAM-Gerlach @jnsebgosselin

CAM-Gerlach commented:
@leportella Good question; its actually somewhat non-trivial to figure that out.

So, assuming you have your fixture set up with qtbot similar to the following, for clarity:

```python
@pytest.fixture
def shortcutssummary_example(qtbot):
    """Set up shortcut summary widget."""
    dlg_shortcuts = ShortcutsSummaryDialog(None)
    qtbot.addWidget(dlg_shortcuts)
    dlg_shortcuts.show()
    yield dlg_shortcuts, qtbot
    dlg_shortcuts.close()
```

and have passed it to your test function like so:

```python
def test_shortcutssummary(shortcutssummary_example):
    """Test that shortcut summary widget shows platform-specific key names."""
    dlg_shortcuts, qtbot = shortcutssummary_example
```

You can then check that the dialog exists and is shown to the user:

```python
# Test that the dialog exists and is shown
assert dlg_shortcuts.isVisible()
```

and you can also make sure it closes properly with the standard keystroke (you'll need to import ``from qtpy.QtCore import Qt``), e.g.:

```python
# Test that the dialog is closed properly on Esc keypress
qtbot.keyClick(dlg_shortcuts, Qt.Key_Escape)
assert not dlg_shortcuts.isVisible()
```

So far, very straightforward. However, in between those two you also want to actually test that the shortcuts are present and printed correctly on macOS as well as other platforms. In order to do that, unless there's an easier way to do it that I'm unaware of, you'll need to dive into the widget/layout hierarchy of the shortcut summary dialog and iterate through each to check each one. To make a long story short, it gets pretty tricky. Try/excepts are needed since some positions don't have items (intentionally) and you'll want to make sure to check that you've found the number of shortcuts you expect, along with their values. Here's the basic outline of how to do that:

```python
    expected_shortcuts = 1
    found_shortcuts = 0
    for column_layout in dlg_shortcuts.scroll_widget.layout().children():
        for group_idx in range(column_layout.count()):
            try:
                group_layout = (column_layout.itemAt(group_idx)
                                .widget().layout())
            except AttributeError:  # Since some groups are not present
                continue
            print(group_layout)
            print("Row count: " + str(group_layout.rowCount()))
            for shortcut_idx in range(group_layout.rowCount()):
                print(shortcut_idx)
                try:
                    shortcut_name = (
                        group_layout.itemAtPosition(shortcut_idx, 0)
                        .widget().text())
                    shortcut_keystr = (
                        group_layout.itemAtPosition(shortcut_idx, 1)
                        .widget().text())
                except AttributeError:  # Since some items are not present
                    continue
                if shortcut_name == "Next cursor position":
                    found_shortcuts += 1
                    if sys.platform.startswith('darwin'):
                        assert shortcut_keystr == "Cmd+Option+Right"
                    else:
                        assert shortcut_keystr == "Ctrl+Alt+Right"
                    break
    assert found_shortcuts == expected_shortcuts
```

I picked ``Next cursor position`` as an example since it included both Ctrl/Cmd and Alt/Option, but you'll want to choose a handful of shortcuts with all the different modifiers to ensure they're all handled properly. You might also need to handle Linux and Windows separately if Meta is handled differently between them. To add more shortcuts to test, you can copy the last block or make it into a four loop; its just an example.

Hope this is helpful!

jnsebgosselin commented:
Thank you @CAM-Gerlach for your example.

I think @ccordoba12 now prefer that we pass `qtbot` as argument to every test that needs it instead of returning it in the fixture. So it would be preferable to have this structure instead:

```python
@pytest.fixture
def dlg_shortcuts(qtbot):
    """Set up shortcut summary widget."""
    dlg_shortcuts = ShortcutsSummaryDialog(None)
    qtbot.addWidget(dlg_shortcuts)
    dlg_shortcuts.show()
    yield dlg_shortcuts
    dlg_shortcuts.close()

def test_shortcutssummary(dlg_shortcuts, qtbot):
    """Test that shortcut summary widget shows platform-specific key names."""
    # Test that the dialog exists and is shown
    assert dlg_shortcuts.isVisible()
    # Test that the dialog is closed properly on Esc keypress
    qtbot.keyClick(dlg_shortcuts, Qt.Key_Escape)
    assert not dlg_shortcuts.isVisible()
```

ccordoba12 commented:
> I think @ccordoba12 now prefer that we pass qtbot as argument to every test that needs it instead of returning it in the fixture

Yep, that's what I prefer right now. The thing is it makes easier to read tests because you know what fixtures it uses by reading its arguments.

----

@leportella, I think you have received very good feedback to add a test for this!

CAM-Gerlach commented:
Thanks @ccordoba12 and @jnsebgosselin ; I wasn't sure what was the currently preferred style since I'd seen both, and I actually had it that way at one point before I changed it, heh.

leportella commented:
Hi guys! Indeed this was not trivial to find out! hahahahahah

I added some tests and I added a file called `helpers` to make the code easier to test with less code. Please let me know what your thoughts on this :)

jnsebgosselin commented on spyder/widgets/tests/helpers.py:
```diff
@@ -0,0 +1,137 @@
+
+
+SHORTCUTS = {
+    'Close pane': ('Shift+Ctrl+F4' ,'⇧⌘F4'),
```
I think you will need to use unicode strings for your tests to pass in Python 2.7, like `u'⇧⌘F4'`

jnsebgosselin commented on spyder/widgets/tests/test_shortcutssummary.py:
```diff
@@ -0,0 +1,72 @@
+
+def test_shortcutssummary_texts(dlg_shortcuts, qtbot):
+    """Test that each shortcut has platform-specific key names."""
+    expected_shortcuts = 124
```
I think this should be 125 because the shortcut `Copy` appears two times, one for the Editor and one for the Variable Explorer.

jnsebgosselin replied:
mmmh, I was not correct here. I don't know why `Copy` is found two times at the end sorry.

jnsebgosselin commented on spyder/widgets/tests/helpers.py:
```diff
@@ -0,0 +1,137 @@
+    'Code completion': ('Meta+Space', u'⌃Space'),
+    'Conditional breakpoint': ('Shift+F12', u'⇧F12'),
+    'Copy': ('Ctrl+C', u'⌘C'),
+    'Copy line': ('Shift+Alt+Down', u'⌥⇧↓'),
```
You may have to treat Windows and Linux differently for `Copy line` and `Duplicate line`

See https://github.com/spyder-ide/spyder/blob/master/spyder/config/main.py#L373

ccordoba12 commented:
I don't know if @jnsebgosselin and @CAM-Gerlach agree, but I'd like to be a bit more pragmatic here. The tests you came up with are great Leticia, but I think in `test_shortcutssummary_texts` you could test just a subset of our shortcuts (let's say 10), that contain `Ctrl`, `Enter`, `Shift` `Alt`, `Meta` and a special key (like `PgDown`) to see if our dialog shows them correctly on Windows, Linux and macOS.

This is because a) we are planning a big refactoring for our shortcuts in the short term; b) some shortcuts will change too (due to some work @jnsebgosselin is planning); and c) I'd like to take advantage of your talent and willingness to help us in other, more critical areas.

jnsebgosselin commented:
I agree. Only testing a subset would be enough and it will be easier to maintain.

leportella commented:
@ccordoba12 is this better?

ccordoba12 commented:
I think that's really nice, thanks!

jnsebgosselin commented:
@leportella, could you please update the screenshot of the final results in MacOS in the OP? I am very curious to see how this looks like. Thank you very much for working on this one!

leportella commented:
@jnsebgosselin I just updated the image on the description of the pull request, ok?

jnsebgosselin commented:
@leportella thanks. Mmmmh, I'm surprised. I was expecting to see some ⇧⌃⌘⌥⌦⎋ symbols in there.

CAM-Gerlach commented:
However, based on the screenshot this is still not functioning correctly, since (as I stated in my review previously), ``Meta`` is not getting replaced with ``Control`` on macOS (most users, especially less experienced ones that platform are unlikely to know what key it corresponds to) as well as ``Backspace`` is not replaced with ``Delete`` (less critical).

CAM-Gerlach commented:
> Sorry for the quick merge. I thought everything was ok.

No worries, I just wasn't getting notifications to see it in a reasonable timeframe.

> This is not so much of a problem since we're going to remove all Meta shortcuts for Windows 10 compatibility, right @jnsebgosselin?

Okay, fair enough.

> Is there no Delete on macOS?

At least back when I was a more regular Mac user, the Backspace key was called delete.

> @leportella, could you expand your test in another PR? Thanks!

In fact, the test appears to not be functioning at all, since its testing for the appearance of the shortcut *symbols* (as they actually appear in macOS menus) which clearly none of which are present according to the screenshots, but they are still passing. I can't test if this works locally since I don't have macOS.

And not that its a big deal, but the post didn't include the developer COI nor affirmation for some reason.

jnsebgosselin commented:
Maybe something went wrong with the update of the screenshot in the OP of the PR. At least, the modification doesn't show in the log.

CAM-Gerlach commented:
Yeah, I figure that's probably what happened—that seems a much more likely explanation. I though that might be the case but I saw the last mod date as 6 hours ago, right around when she made her post, and for whatever reason didn't actually check the full log.

leportella commented:
@jnsebgosselin and @CAM-Gerlach I am so sorry for the confusion. I just showed @ccordoba12 that the screenshot was not updated, and everything is working correctly. I tested on the master branch and the result is this:

<img width="1062" alt="screen shot 2018-11-19 at 13 03 05" src="https://user-images.githubusercontent.com/7255297/48717245-26ce2780-ec00-11e8-8299-c501f2d72244.png">

jnsebgosselin commented:
That is great, thank you very much!

ccordoba12 commented:
Really nice!! Thanks a lot Leticia!

CAM-Gerlach commented:
Thanks @leportella ! I updated the out of date screenshot in your OP with that one.
